scipy/_lib/_util.py

```diff
@@ -337,3 +338,69 @@ def getargspec_no_self(func):
         if argspec.args[0] == 'self':
             argspec.args.pop(0)
         return argspec
+
+
+class PoolWrapper(object):
+    """
+    Wrapper for working with objects that have map-like methods, such as
+    `multiprocessing.Pool`. Used for parallelisation.
+
+    Parameters
+    ----------
+    pool : int or map-like object
+        If `pool` is an `int` then it specifies the number of threads to
+        use for parallelization. If `int(pool) in [0, 1]`, then no parallel
+        processing is used and the map builtin is used. If `pool == -1` then
+        the pool will utilise all available CPU.
+        If pool is an object with a map method that follows the same
+        calling sequence as the built-in map function, then this pool is
+        used for parallelisation.
+    """
+    def __init__(self, pool=-1):
+        self.pool = None
+        self._mapfunc = map
+        self._own_pool = False
+
+        if hasattr(pool, 'map'):
+            self.pool = pool
+            self._mapfunc = self.pool.map
+        else:
+            # user supplies a number
+            if int(pool) == -1:
+                # use as many processors as possible
+                self.pool = Pool()
+                self._mapfunc = self.pool.map
+                self._own_pool = True
+            elif int(pool) in [0, 1]:
+                pass
+            elif int(pool) > 1:
+                # use the number of processors requested
+                self.pool = Pool(processes=int(pool))
+                self._mapfunc = self.pool.map
+                self._own_pool = True
+
+    def __enter__(self):
+        return self
+
+    def __del__(self):
+        self.close()
+
+    def terminate(self):
+        if self._own_pool:
+            self.pool.terminate()
+
+    def join(self):
+        if self._own_pool:
+            self.pool.join()
+
+    def close(self):
+        if self._own_pool:
+            self.pool.close()
+
+    def __exit__(self, *args):
+        if self._own_pool:
+            self.pool.terminate()
```

Comment:
This should be:
```
def __exit__(self, exc_type, exc_value, traceback):
    if self._own_pool:
        if exc_type is None:
            self.pool.close()
            self.pool.join()
        else:
            self.pool.terminate()
```
Calling terminate sends signals around killing processes, and usually should not be used for "successful" exit.

Reply:
Done